File: examples/blazedGratingsEtching/blazedGratingsEtching.cpp
#include "blazedGratingsGeometry.hpp"

#include <models/psIonBeamEtching.hpp>
#include <process/psProcess.hpp>
#include <psConstants.hpp>

using namespace viennaps;

int main(int argc, char **argv) {
  omp_set_num_threads(16);
  Logger::setLogLevel(LogLevel::DEBUG);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    // Try default config file
    params.readConfigFile("config.txt");
    if (params.m.empty()) {
      std::cout << "No configuration file provided!" << std::endl;
      std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
      return 1;
    }
  }

  auto geometry =
      GenerateMask<double, 2>(params.get("bumpWidth"), params.get("bumpHeight"),
                              params.get<int>("numBumps"),
                              params.get("bumpDuty"), params.get("gridDelta"));
  geometry->saveSurfaceMesh("initial");

  AdvectionParameters advParams;
  advParams.spatialScheme =
      viennals::SpatialSchemeEnum::LAX_FRIEDRICHS_2ND_ORDER;
  advParams.timeStepRatio = 0.25;

  RayTracingParameters rayTracingParams;
  rayTracingParams.raysPerPoint = params.get<unsigned>("raysPerPoint");
  rayTracingParams.smoothingNeighbors = 1;

  const double yieldFac = params.get("yieldFactor");

  IBEParameters<double> ibeParams;
  ibeParams.materialPlaneWaferRate[Material::SiO2] = 1.0;
  ibeParams.materialPlaneWaferRate[Material::Mask] = 1. / 11.;
  ibeParams.exponent = params.get("exponent");
  ibeParams.meanEnergy = params.get("meanEnergy");
  ibeParams.cos4Yield.isDefined = true;
  ibeParams.cos4Yield.a1 = yieldFac;
  ibeParams.cos4Yield.a2 = -1.55;
  ibeParams.cos4Yield.a3 = 0.65;

  // ANSGM Etch
  {
    double angle = params.get("phi1");
    double angleRad = constants::degToRad(angle);
    ibeParams.tiltAngle = angle;
    auto model = SmartPointer<IonBeamEtching_double_2>::New(ibeParams);
    model->setPrimaryDirection(
        Vec3Dd{-std::sin(angleRad), -std::cos(angleRad), 0});
    model->setProcessName("ANSGM_Etch");

    Process_double_2(geometry, model, params.get("ANSGM_Depth"), advParams,
                     rayTracingParams)
        .apply();
    geometry->saveSurfaceMesh("ANSGM_Etch");
  }

  geometry->removeTopLevelSet(); // remove mask
  geometry->saveSurfaceMesh("ANSGM");

  // Blazed Gratings Etch
  {
    double angle = params.get("phi2");
    double angleRad = constants::degToRad(angle);
    ibeParams.tiltAngle = angle;
    auto model = SmartPointer<IonBeamEtching_double_2>::New(ibeParams);
    model->setPrimaryDirection(
        Vec3Dd{-std::sin(angleRad), -std::cos(angleRad), 0});
    model->setProcessName("BlazedGratings_Etch");

    Process_double_2 process(geometry, model);
    process.setParameters(advParams);
    process.setParameters(rayTracingParams);

    for (int i = 1; i < 5; ++i) {
      process.setProcessDuration(params.get("etchTimeP" + std::to_string(i)));
      process.apply();
      geometry->saveSurfaceMesh("BlazedGratingsEtch_P" + std::to_string(i));
    }
  }
}

File: examples/blazedGratingsEtching/blazedGratingsEtching.py
import viennaps as ps
import numpy as np
from argparse import ArgumentParser

# parse config file name and simulation dimension
parser = ArgumentParser(
    prog="blazedGratingsEtching",
    description="Run a oblique ion beam etching process on a grating.",
)
parser.add_argument("filename")
args = parser.parse_args()

params = ps.readConfigFile(args.filename)

# ----- Geometry Generation ----- #
bumpWidth = params["bumpWidth"]
bumpHeight = params["bumpHeight"]
bumpDuty = params["bumpDuty"]
numBumps = int(params["numBumps"])
bumpSpacing = bumpWidth * (1.0 - bumpDuty) / bumpDuty
xExtent = numBumps * bumpWidth / bumpDuty

geometry = ps.Domain(
    gridDelta=params["gridDelta"],
    xExtent=xExtent,
    boundary=ps.BoundaryType.PERIODIC_BOUNDARY,
)
ps.MakePlane(domain=geometry, height=0.0, material=ps.Material.SiO2).apply()
mask = ps.ls.Domain(geometry.getGrid())

mesh = ps.ls.Mesh()
offset = -xExtent / 2.0 + bumpSpacing + bumpWidth / 2.0
numNodes = 100
for i in range(numNodes):
    x = -bumpWidth / 2.0 + i * bumpWidth / (numNodes - 1)
    y = -4.0 * bumpHeight * x * x / (bumpWidth * bumpWidth) + bumpHeight
    mesh.insertNextNode([x + offset, y, 0.0])
for i in range(1, numNodes):
    mesh.insertNextLine([i - 1, i])
mesh.insertNextLine([numNodes - 1, 0])

for i in range(numBumps):
    tip = ps.ls.Domain(geometry.getGrid())
    ps.ls.FromSurfaceMesh(tip, mesh).apply()
    ps.ls.TransformMesh(
        mesh=mesh,
        transform=ps.ls.TransformEnum.TRANSLATION,
        transformVector=[bumpSpacing + bumpWidth, 0, 0],
    ).apply()
    ps.ls.BooleanOperation(mask, tip, ps.ls.BooleanOperationEnum.UNION).apply()

geometry.insertNextLevelSetAsMaterial(mask, ps.Material.Mask)
geometry.saveSurfaceMesh("initial", True)

# ----- Model Setup ----- #
advectionParams = ps.AdvectionParameters()
advectionParams.spatialScheme = ps.SpatialScheme.LAX_FRIEDRICHS_2ND_ORDER
advectionParams.timeStepRatio = 0.25

rayTracingParams = ps.RayTracingParameters()
rayTracingParams.raysPerPoint = int(params["raysPerPoint"])
rayTracingParams.smoothingNeighbors = 1

ibeParams = ps.IBEParameters()
ibeParams.exponent = params["exponent"]
ibeParams.meanEnergy = params["meanEnergy"]
ibeParams.materialPlaneWaferRate = {ps.Material.SiO2: 1, ps.Material.Mask: 1 / 11}
ibeParams.cos4Yield.isDefined = True
ibeParams.cos4Yield.a1 = params["yieldFactor"]
ibeParams.cos4Yield.a2 = -1.55
ibeParams.cos4Yield.a3 = 0.65

model = ps.IonBeamEtching()

# ----- ANSGM Etch ----- #
angle = params["phi1"]
direction = [0.0, 0.0, 0.0]
direction[0] = -np.sin(np.deg2rad(angle))
direction[1] = -np.cos(np.deg2rad(angle))
ibeParams.tiltAngle = angle
model.setPrimaryDirection(direction)
model.setParameters(ibeParams)

process = ps.Process(geometry, model, 0.0)
process.setParameters(advectionParams)
process.setParameters(rayTracingParams)

process.setProcessDuration(params["ANSGM_Depth"])
process.apply()
geometry.saveSurfaceMesh("ANSGM_Etch", True)

# remove mask
geometry.removeTopLevelSet()
geometry.saveSurfaceMesh("ANSGM", True)

# ------ Blazed Gratins Etch ------ #
angle = params["phi2"]
direction[0] = -np.sin(np.deg2rad(angle))
direction[1] = -np.cos(np.deg2rad(angle))
ibeParams.tiltAngle = angle
model.setPrimaryDirection(direction)
model.setParameters(ibeParams)

for i in range(1, 5):
    process.setProcessDuration(params["etchTimeP" + str(i)])
    process.apply()
    geometry.saveSurfaceMesh("BlazedGratingsEtch_P" + str(i))
